torch/_inductor/codegen/triton.py

```diff
@@ -1050,45 +1056,70 @@ def mask_loads(self, mask):
-
-        indirect_vars = [
-            s for s in original_index.free_symbols if s.name.startswith("tmp")
-        ]
-        for var in indirect_vars:
-            sizes = list(var_size[var])
-            if all(isinstance(s, sympy.Integer) for s in sizes):
-                size = min(sizes)
-            else:
-                # Should this go here or in TritonPrinter?
-                def print_min(expr):
-                    if len(expr) == 1:
-                        return texpr(expr[0])
-                    else:
-                        return f"min({texpr(expr[0])}, {print_min(expr[1:])})"
-
-                size = print_min(sizes)
-            # The conditions need to be in parens because of Python's operator precedence.
-            # It'd be less # error-prone to use and/or/not, which is suported by triton
-            cond = f"((0 <= {var}) & ({var} < {size}))"
-            cond_print = f"0 <= {var} < {size}"
-            if not isinstance(original_index, sympy.Integer):
-                var_mask = f"({mask})" if "&" in mask else mask
-                var_mask = f" | ~{var_mask}"
+    def indirect_indexing(self, var, size):
+        class IndirectAssertLine(DeferredLineBase):
+            def __init__(self, line, var, mask, size_map):
+                self.var = var
+                self.mask = mask
+                self.line = line
+                self.size_map = size_map
+
+            def __call__(self):
+                # The conditions need to be in parens because of Python's operator precedence.
+                # It'd be less # error-prone to use and/or/not, which is suported by triton
+                size = self.size_map[(self.var, self.mask)]
+                cond = f"(0 <= {self.var}) & ({self.var} < {size})"
+                cond_print = f"0 <= {self.var} < {size}"
+                if self.mask:
+                    cond = f"({cond}) | ~{self.mask}"
+                return self.line.format(cond=cond, cond_print=cond_print)
+
+            def _new_line(self, line):
+                return IndirectAssertLine(line, self.var, self.mask, self.size_map)
+
+        var_str = str(var)
+
+        if config.triton.assert_indirect_indexing and torch.version.hip is None:
+            mask_vars = set(var.mask_vars)
+            if self._load_mask:
+                mask_vars.add(self._load_mask)
+
+            mask = ""
+            if mask_vars:
+                mask = (
+                    f"{list(mask_vars)[0]}"
+                    if len(mask_vars) == 1
+                    else f"({' & '.join(str(v) for v in mask_vars)})"
+                )
+
+            # tl.device_assert doesn't work for constexpr values, and we can't
+            # tell from here if a var is constexpr or not, so promote everything
+            var_str = str(
+                self.cse.generate(
+                    self.compute, f"triton_helpers.promote_to_tensor({var})"
+                )
+            )
```

Comment:
This is not going to do as per https://github.com/pytorch/pytorch/pull/100880#issuecomment-1540494933. Now, I think that most the "indirect indexing via scalars" (aka direct indexing) we do is not tagged as indirect_indexing, but I guess that there are some times where this is not the case and we do things like:
```python
tmp0 = 0
tl.load(in_ptr0 + tmp0, ...)
```
I don't know when is it that we generate this sort of code though.